### ISSUE TITLE
Segmentation fault in maild when 'smtp_server' change position in 'ossec.conf'

### ISSUE THREAD

**Issue by DaveVG1** [wazuh/wazuh#2688]
|Wazuh version|Install type|Install method|
|---|---|---|
| 3.9.0 | Manager/Agent | Packages/Sources |

Bug description
---
If default configuration is used, `maild` works as usual, but if `smtp_server` is defined above of `email_notification` a segmentation fault is produced. This also happens with the configuration of `helo_server`.

This happens because `Mail->mn` (`email_notification`) is not set at the moment we read `smtp_server` (because it's above in the configuration), leaving it unset.
https://github.com/wazuh/wazuh/blob/1935c4b203fb568ea60ac635b64c32b7705d8544/src/config/global-config.c#L517
After that, we try to read `smpt_server`, causing the segmentation fault:
https://github.com/wazuh/wazuh/blob/1935c4b203fb568ea60ac635b64c32b7705d8544/src/os_maild/maild.c#L164

**Comment by juan70** [comment]
2019-02-26
----------
  - Debugging and finding a solution.

**Comment by juan70** [comment]
Source code involved:
`src/os_maild_maild.c`
```
121     /* Read configuration */
122     if (MailConf(test_config, cfg, &mail) < 0) {
123         merror_exit(CONFIG_ERROR, cfg);
124     }
(...)
164     if (mail.smtpserver[0] != '/') {
(...)
```

`src/os_maild/config.c`
```
30     Mail->mn = 0;
```

`src/config/global-config.c`
```
221         else if (strcmp(node[i]->element, xml_mailnotify) == 0) {
222             if (strcmp(node[i]->content, "yes") == 0) {
223                 if (Config) {
224                     Config->mailnotify = 1;
225                 }
226                 if (Mail) {
227                     Mail->mn = 1;
228                 }
229             } else if (strcmp(node[i]->content, "no") == 0) {
230                 if (Config) {
231                     Config->mailnotify = 0;
232                 }
233                 if (Mail) {
234                     Mail->mn = 0;
235                 }
(...)
515         } else if (strcmp(node[i]->element, xml_smtpserver) == 0) {
516 #ifndef WIN32
517             if (Mail && (Mail->mn)) {
518                 if (node[i]->content[0] == '/') {
519                     os_strdup(node[i]->content, Mail->smtpserver);
520                 } else {
521                     Mail->smtpserver = OS_GetHost(node[i]->content, 5);
522                     if (!Mail->smtpserver) {
523                         merror(INVALID_SMTP, node[i]->content);
524                         return (OS_INVALID);
525                     }
526                 }
527             }
```

With this configuration in `ossec.conf`:
```
  <global>
    (...)
    <smtp_server>smtp.gmail.com</smtp_server>
    <email_notification>yes</email_notification>
    (...)
  </global>
```
A segmentation fault is produced when `ossec-maild` starts.

In order for `Mail->smtpserver` to contain a significant value, `Mail->mn` must be `1` (`= yes`).
By default, `Mail->mn` is `0`. If `<email_notification>yes` is placed after `<smtp_server>`, the value of `Mail->smtpserver` will not be set.
After that, `Mail->mn` will be set to `1`. Later, in `if (mail.smtpserver[0] != '/') {`, the variable `mail.smtpserver` is `NULL`, hence the segmentation fault.
It `Mail->mn` is still `0`, because `<email_notification>no`, the `ossec-maild` process will not be started, no attempt to read `mail.smtpserver` will be performed, and no segmentation will never occur with that configuration.

Possible solution:
- When reading `<smtp_server>`, if `Mail->mn == 0`, make a temporary copy of the contents of the node. Same thing for`<helo_server>`.
- When reading `<email_notification>` and `Mail->mn` is set to `1`, then set the server value using the temporary value (if it is set because read earlier).

**Comment by juan70** [comment]
# Testing (again)
With <email_notification>no (either explicitely or by default)
--
- Fresh manager install, `<email_notification>` is initially configured with `no`, and `<smtp_server>` contains an example.
    ```
    <email_notification>no</email_notification>
    <smtp_server>smtp.example.wazuh.com</smtp_server>
    ```
    => no error, maild does not start, manager starts => OK
  Change the order:
    ```
    <smtp_server>smtp.example.wazuh.com</smtp_server>
    <email_notification>no</email_notification>
    ```
    => no error, maild does not start, manager starts => OK

- With no tags related to email_notification (no `<email_notification>`, no `<smtp_server>`, and no `<helo_server>`)
    => no error, maild does not start, manager starts => OK

- Only `<email_notification>no` (i.e. no `<smtp_server>` or `<helo_server>`)
    => no error, maild does not start, manager starts => OK

- Only `<helo_server>` (by default, email_notification=no)
    => no error, maild does not start, manager starts => OK

- Only `<smtp_server>` (by default, email_notification=no)
    => no error, maild does not start, manager starts => OK


With <email_notification>yes
--
- Only `<email_notification>yes` (no `<helo_server>` or `<smtp_server`>)
    => no error message, maild does not start, manager starts => FAIL: error message expected because SMTP server not set


- `<email_notification>yes` + `<helo_server></helo_server>` (in this order)
    That is
    ```
        <email_notification>yes</email_notification>
        <helo_server></helo_server>
    ```
    => no error message, maild does not start, manager starts => FAIL: error message expected because SMTP server not set

  The other way around, that is:
    ```
        <helo_server></helo_server>
        <email_notification>yes</email_notification>
    ```
    => no error message, maild does not start, manager starts => FAIL: error message expected because SMTP server not set

- `<email_notification>yes` + `<helo_server>/some/path</helo_server>`
    => no error message, maild does not start, manager starts => FAIL: error message expected because SMTP server not set
  The other way around
    => no error message, maild does not start, manager starts => FAIL: error message expected because SMTP server not set

- `<email_notification>yes` + `<smtp_server></smtp_server>`
    => no error message, maild does not start, manager starts => FAIL: error message expected because SMTP server not set
  The other way around
    => no error message, maild does not start, manager starts => FAIL: error message expected because SMTP server not set

- `<email_notification>yes` + `<smtp_server>/some/path</smtp_server>`
    => manager starts, maild does not starts, no error => FAIL, maild should start?
  The other way around
    => manager starts, maild does not starts, no error => FAIL, maild should start?

- `<email_notification>yes` + `<smtp_server>smtp.gmail.com</smtp_server>`
    => manager starts, maild does not starts, no error => FAIL, maild should start
  The other way around
    => manager starts, maild does not starts, no error => FAIL, maild should start

- `<email_notification>yes` + `<smtp_server>some_random_text</smtp_server>`
    => manager starts, no error => FAIL, expected error 'ERROR: Invalid SMTP Server' 
  The other way around
    => manager starts, no error => FAIL, expected error 'ERROR: Invalid SMTP Server' 


But, with a fresh install, if `ossec.conf` is changed immediately:
```
        <email_notification>yes</email_notification>
        <smtp_server>smtp.gmail.com</smtp_server>
```
The manager starts, maild starts, no error. OK. (Actually, I get an error, but because the mail server is not configured, and no mail can be sent. So, that's normal.)

Summary (kind of)
--
Something weird seems to be happening during the tests, maybe in the transition from `<email_notification>no` to `<email_notification>yes`, maybe combined with `<smtp_server></smtpserver>` (empty tag). Sometimes `maild` starts, sometimes it does not, even though the configuration seems correct.
When `maild` does not start, the exit occurs in: https://github.com/wazuh/wazuh/blob/161387eef6fdc7444959f6a3fb51817e4deb7c9f/src/os_maild/maild.c#L150
Sometimes, `maild.c` is executed twice: first time with `test_config = 1`, and second time with `test_config = 0`. Sometimes, the case `test_config = 0` does not happen. If this is caused by an error somewhere, there is no error message describing what may have happen

**Comment by chemamartinez** [comment]
Merged: #2711
